README.md
Deze app werd ontwikkeld om mensen met een beperking te helpen bij het cash betalen in de winkel.
Ze kunnen eenvoudiger zien wat ze moeten betalen en wat ze (eventueel) terug moeten krijgen.


CHANGELOG.md
## Next version (Unreleased)

## 0.8 (15-04-2019)

BREAKING CHANGES:

- Progressive Web App: moneycalc is now a PWA to work offline

Previous

The changelog began with version 0.7


scripts/menu.js
$(document).ready(function() {
    function closeMenu() {
        $("#menu").animate({"left": "-=250px"}, "normal", function() {
            $(this).removeClass("out");
        });
        $("#lnk-mobile-menu").animate({left: "-=250"}, 400);
        $("#wrapper").animate({left: "-=250"}, 400).css("position","relative");
    }

    $("#lnk-mobile-menu").click(function(e) {
        e.stopPropagation();
		if($("#menu").hasClass("out")) {
            closeMenu();
		} else {
			$("#menu").addClass("out");
			$("#menu").animate({left: "+=250"}, 400);
			$("#lnk-mobile-menu").animate({left: "+=250"}, 400);
			$("#wrapper").animate({left: "+=250"}, 400).css("position","fixed");
		}
        return false;
	});
    
    $("#wrapper").click(function() {
        if($("#menu").hasClass("out")) {
            closeMenu();
        }
    });
    
    $(window).on('resize', function() {
        if($(window).width() > 768) {
            if($("#menu").hasClass("out")) {
                closeMenu();           
            }
        }
    });    
});


scripts/modal.js
window.addEventListener('load', function() {
    if(localStorage.getItem("show-update-info") == null) {
        document.getElementById("update-info").classList.add("open");
    }
})

document.addEventListener('click', function (e) {
    e = e || window.event;
    var target = e.target || e.srcElement;
    
    if ((target.hasAttribute('data-dismiss') && target.getAttribute('data-dismiss') == 'modal') || target.classList.contains('modal')) {
        var modal = document.querySelector('[class="modal open"]');
        modal.classList.remove('open');
        e.preventDefault();
    } else if ((target.hasAttribute('data-dismiss') && target.getAttribute('data-dismiss') == 'modal-button') || target.classList.contains('modal')) {
        var modal = document.querySelector('[class="modal open"]');
        modal.classList.remove('open');
        localStorage.setItem("show-update-info", 0);
        e.preventDefault();
    }
}, false);

service-worker.js
var dataCacheName = 'moneycalc-v1';
var cacheName = 'moneycalc-v1';
var filesToCache = [
  'https://moneycalc.bartleemans.be/',
  'https://moneycalc.bartleemans.be/index.html',
  'https://moneycalc.bartleemans.be/betalen.html',
  'https://moneycalc.bartleemans.be/info.html',
  'https://moneycalc.bartleemans.be/rekenmachine.html',
  'https://moneycalc.bartleemans.be/scripts/errorhandler.js',
  'https://moneycalc.bartleemans.be/scripts/jquery-1.10.1.min.js',
  'https://moneycalc.bartleemans.be/scripts/menu.js',
  'https://moneycalc.bartleemans.be/scripts/init.js',
  'https://moneycalc.bartleemans.be/scripts/main.js',  
  'https://moneycalc.bartleemans.be/css/menu.css',
  'https://moneycalc.bartleemans.be/css/modal.css',
  'https://moneycalc.bartleemans.be/css/main.css',
  'https://moneycalc.bartleemans.be/favicon.ico',
  'https://moneycalc.bartleemans.be/images/icons/icon-16x16.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-32x32.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-128x128.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-144x144.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-152x152.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-192x192.png',
  'https://moneycalc.bartleemans.be/images/icons/icon-256x256.png',
  'https://moneycalc.bartleemans.be/images/icons/android-chrome-192x192.png',
  'https://moneycalc.bartleemans.be/images/icons/android-chrome-512x512.png',
  'https://moneycalc.bartleemans.be/images/icons/apple-touch-icon.png',
  'https://moneycalc.bartleemans.be/images/icons/mstile-150x150.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-640x1136.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-750x1294.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-1125x2436.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-1242x2148.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-1536x2048.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-1668x2224.png',
  'https://moneycalc.bartleemans.be/images/splash/launch-2048x2732.png',
  'https://moneycalc.bartleemans.be/images/money/1.png',
  'https://moneycalc.bartleemans.be/images/money/2.png',
  'https://moneycalc.bartleemans.be/images/money/5.png',
  'https://moneycalc.bartleemans.be/images/money/10.png',
  'https://moneycalc.bartleemans.be/images/money/20.png',
  'https://moneycalc.bartleemans.be/images/money/50.png',
  'https://moneycalc.bartleemans.be/images/money/100.png',
  'https://moneycalc.bartleemans.be/images/money/200.png',
  'https://moneycalc.bartleemans.be/images/money/500.png',
  'https://moneycalc.bartleemans.be/images/money/1000.png',
  'https://moneycalc.bartleemans.be/images/money/2000.png',
  'https://moneycalc.bartleemans.be/images/money/5000.png',
  'https://moneycalc.bartleemans.be/images/money/10000.png',
  'https://moneycalc.bartleemans.be/images/money/20000.png'
];

self.addEventListener('install', function(e) {
  e.waitUntil(
    caches.open(cacheName).then(function(cache) {
      return cache.addAll(filesToCache);
    })
  );
});

self.addEventListener('activate', function(e) {
  e.waitUntil(
    caches.keys().then(function(keyList) {
      return Promise.all(keyList.map(function(key) {
        if(key !== cacheName && key !== dataCacheName) {
          return caches.delete(key);
        }
      }));
    })
  );
  return self.clients.claim();
});

self.addEventListener('fetch', function(event) {
  event.respondWith(
    caches.match(event.request).then(function(response) {
      if (response) {
        return response;
      } else {
        return fetch(event.request).then(function(res) {
          return caches.open('dynamic').then(function(cache) {
            cache.put(event.request.url, res.clone());
            return res;
          });
        });
      }
    })
  );
});


scripts/main.js
var app = {
	isLoading: true,
	version: "01082018v1",
	originalAmount: 0,
	moneyToPay: undefined,
	getMoney: function() {		
		var a = this.getResult("1");
		var b = this.getResult("2");
		var c = this.getResult("5");
		var d = this.getResult("10");
		var e = this.getResult("20");
		var f = this.getResult("50");
		var g = this.getResult("100");
		var h = this.getResult("200");
		var i = this.getResult("500");
		var j = this.getResult("1000");
		var k = this.getResult("2000");
		var l = this.getResult("5000");
		var m = this.getResult("10000");
	
		var money = [[1, a], [2, b], [5, c], [10, d], [20, e], [50, f], [100, g], [200, h], [500, i], [1000, j], [2000, k], [5000, l], [10000, m]];

		return money;
	},
	init: function() {		
		app.initStorage();
		app.calculateTotal();
	},
	initStorage: function() {
		var money = this.getMoney();
		//init coins and bills
		for(i = 0; i <= 12; i++) {
			if(money[i][1] == null) {
				app.storeResult(money[i][0], 0);
			}
		}
	},
	resetStorage: function() {
		var money = this.getMoney();
		for(i = 0; i <= 12; i++) {
			app.storeResult(money[i][0], 0);
		}
	},
	calculateTotal: function() {
		var money = this.getMoney();		
		var total = (parseInt(money[0][1]) * 1) + (parseInt(money[1][1]) * 2) + (parseInt(money[2][1]) * 5) + (parseInt(money[3][1]) * 10) + (parseInt(money[4][1]) * 20) + (parseInt(money[5][1]) * 50) + (parseInt(money[6][1]) * 100) + (parseInt(money[7][1]) * 200) + (parseInt(money[8][1]) * 500) + (parseInt(money[9][1]) * 1000) + (parseInt(money[10][1]) * 2000) + (parseInt(money[11][1]) * 5000) + (parseInt(money[12][1]) * 10000);		
		document.getElementById("total").innerHTML = (total / 100).toFixed(2) + " euro";
		for(i = 0; i <= 12; i++) {
			var id = "valuta-" + money[i][0];
			document.getElementById(id).innerHTML = money[i][1] + " x";
		}
	},
	add: function(id) {
		if(typeof(Storage) !== "undefined") {
			var currVal = parseInt(localStorage.getItem(id));
			localStorage.setItem(id, currVal + 1);
			this.calculateTotal();
		}
	},
	remove: function(id) {
		if(typeof(Storage) !== "undefined") {
			var currVal = parseInt(localStorage.getItem(id));
			var newVal = currVal - 1;
			if(newVal < 0) {
				newVal = 0;
			}
			localStorage.setItem(id, newVal);
			this.calculateTotal();
		}
	},
	calculatePayment: function(amount) {
		if(amount == undefined) {
			//convert to cents for easy calculation
			amount = document.getElementById("payment-amount").value.replace(",", ".");
			amount = (eval(amount).toFixed(2)).replace(".","");
			app.originalAmount = amount;
		}
		var output = document.getElementById("output");
		var money = this.getMoney();
		var total = (parseInt(money[0][1]) * 1) + (parseInt(money[1][1]) * 2) + (parseInt(money[2][1]) * 5) + (parseInt(money[3][1]) * 10) + (parseInt(money[4][1]) * 20) + (parseInt(money[5][1]) * 50) + (parseInt(money[6][1]) * 100) + (parseInt(money[7][1]) * 200) + (parseInt(money[8][1]) * 500) + (parseInt(money[9][1]) * 1000) + (parseInt(money[10][1]) * 2000) + (parseInt(money[11][1]) * 5000) + (parseInt(money[12][1]) * 10000);
		if(amount <= total) {
			var message = "";
			var visualMessage = "Betaal met: <br>";
			var changeMoney = "";

			//check if amount exactly available
			if(amount > 0){
				for(i = 0; i < money.length - 1; i++) {
					for(j = money[i][1] - 1; j >= 0; j--) {
						message = message + app.convertToEuro(money[i][0]) + ",";
						visualMessage = visualMessage + '<img src="images/money/' + money[i][0] + '.png">';
						amount = amount - money[i][0];
						money[i][1]--;
						if(amount <= 0) {
							break;
						}
					}
				}
			} else {
				message = "saldo ontoereikend";
			}

			if(amount != 0){
				money = this.getMoney();
				amount = app.originalAmount;
				message = "";
				visualMessage = "";

				for(i = (money.length - 1); i >=0; i--) {
					if(money[i][0] <= amount) {						
						if(money[i][1] > 0) {						
							for(j = money[i][1] - 1; j >= 0; j--) {
								message = message + app.convertToEuro(money[i][0]) + ",";
								visualMessage = visualMessage + '<img src="images/money/' + money[i][0] + '.png">';
								amount = amount - money[i][0];
								money[i][1]--;
								if(amount <= 0 ) {
									changeMoney = amount;
									break;
								}
							}
						}
					}
				}
			} else {
				message = "saldo ontoereikend";
			}
			
			//als er nog een bedrag over is dan moeten we een hoger item zoeken	en dus voorgaande resetten
			if(amount > 0) {
				money = this.getMoney();
				amount = app.originalAmount;
				message = "";
				visualMessage = "Betaal met: <br>";
				var changeMoney = app.originalAmount;				
				//changemoney nog uitwerken
				for(i = 0; i <= (money.length - 1); i++) {
					if(money[i][0] >= amount && money[i][1] > 0) {
						message = message + app.convertToEuro(money[i][0]) + ",";
						visualMessage = visualMessage + '<img src="images/money/' + money[i][0] + '.png">';					
						money[i][1]--;
						amount = amount - money[i][0];
						changeMoney = amount;
						if(money <= 0) {
							break;
						}
					}
				}
			}
			app.moneyToPay = money;
			output.innerHTML = '<br>' + visualMessage + '<br><br>Terugkrijgen: ' + (Math.abs(changeMoney) / 100) + "&euro;";
		} else {
			output.innerHTML = '<span class="error">Onvoldoende budget</span>';
		}
	},
	confirmPayment() {
		if(app.moneyToPay != undefined) {
			for(i = (app.moneyToPay.length - 1); i >=0; i--) {
				localStorage.setItem(app.moneyToPay[i][0], app.moneyToPay[i][1]);
			}
			app.moneyToPay = undefined;
		}
		window.location = "index.html";
	},
	storeResult(name, val) {
		if(typeof(Storage) !== "undefined") {
			localStorage.setItem(name, val);
		}
	},
	getResult(id) {
		if(typeof(Storage) !== "undefined") {
			return localStorage.getItem(id);
		}
	},
	convertToEuro($cents) {
		return $cents / 100 + " euro";
	}
};

var calculator = {
	action: "",
	calculation: "",
	pos: 0,
	setCalculatorValue: function(val, action) {
		var fv = document.getElementById("result");
		if(action == "=") {
			var regex1 = new RegExp(':', 'g');
			var regex2 = new RegExp('x', 'g');
			calculator.checkPostfixes();
			fv.value = calculator.calculation + "=" + (eval(calculator.calculation.replace(regex1, "/").replace(regex2, "*")).toFixed(2));
			calculator.calculation = "";
		} else if((action == '+') || (action == '-') || (action == 'x') || (action == ':')) {				
			calculator.checkPostfixes();
			calculator.calculation = calculator.calculation + val;
			fv.value = calculator.calculation;
		} else if(action == "clear") {
			calculator.reset();
		} else {
			calculator.calculation = calculator.calculation + val;
			fv.value = calculator.calculation;
		}
		calculator.pos = calculator.pos + 1;
	},
	reset: function() {
		var fv = document.getElementById("result");
		fv.value = calculator.calculation;
		calculator.calculation = "";
		calculator.pos = 0;
	},
	checkPostfixes: function() {
		if(calculator.calculation.charAt(calculator.pos - 1) == ".") {
			calculator.calculation = calculator.calculation + "00";
			calculator.pos = calculator.pos + 2;
		} else if(calculator.calculation.charAt(calculator.pos - 2) == ".") {
			calculator.calculation = calculator.calculation + "0";
			calculator.pos = calculator.pos + 1;
		} else if(calculator.calculation.charAt(calculator.pos - 3) != ".") {
			calculator.calculation = calculator.calculation + ".00";
			calculator.pos = calculator.pos + 3;
		}
	}
};